firedrake/eigensolver.py

```diff
@@ -0,0 +1,187 @@
+"""Specify and solve finite element eigenproblems."""
+from firedrake.assemble import assemble
+from firedrake.function import Function
+from firedrake import utils
+from firedrake.petsc import OptionsManager, flatten_parameters
+from firedrake.exceptions import ConvergenceError
+try:
+    from slepc4py import SLEPc
+except ImportError:
+    SLEPc = None
+__all__ = ["LinearEigenproblem",
+           "LinearEigensolver"]
+
+
+class LinearEigenproblem():
+    """Generalised linear eigenvalue problem
+
+    The problem has the form::
+
+        A(u, v) = λ * M(u, v)
+
+    Parameters
+    ----------
+    A : ufl.Form
+        The bilinear form A(u, v).
+    M : ufl.Form
+        The mass form M(u, v),  defaults to u * v * dx.
+    bcs : DirichletBC or list of DirichletBC
+        The boundary conditions.
+    bc_shift: float
+        The value to shift the boundary condition eigenvalues by.
+
+    Notes
+    -----
+
+    If Dirichlet boundary conditions are supplied then these will result in the
+    eigenproblem having a nullspace spanned by the basis functions with support
+    on the boundary. To facilitate solution, this is shifted by the specified
+    amount. It is the user's responsibility to ensure that the shift is not
+    close to an actual eigenvalue of the system.
+    """
+    def __init__(self, A, M=None, bcs=None, bc_shift=666.0):
```

Comment:
Why the bc_shift by default, not zero? This is would give an Inf eigenvalue for each bc node.

Reply:
You answered your own question. If `bc_shift` is zero then the boundary eigenvalues are mathematically undefined. Expecting a numerical algorithm to do something sensible in those circumstances is optimistic, to say the least. In particular, the way SLEPc will solve the generalised eigenvalue problem is $M^{-1}Au = \lambda u$. If `bc_shift==0` then $M$ has rows which are zero and the solve $M^{-1}(Au)$ is highly likely to fail. Using a finite shift makes the problem non-singular.

Reply:
Let me be more precise, my understanding is that if the `bc_shift` is zero only on the mass matrix there can not be any boundary eigenvalue because Inf is not a real number.
If a matrix in the pencil of the generalised eigenvalue problem is singular this doesn't mean that the pencil itself is singular.
Indeed SLEPc will solve the generalised eigenvalue problem as $M^{-1}Au=\lambda u$ by default, but my understanding is that this is incorrect if your $M$ is singular (for obvious reasons). In my opinion, the correct thing to do is to use a spectral transformation and my understanding is that the standard choice in this case is the SINVERT spectral transform, i.e. SLEPc will solve the generalised eigenvalue problem as $(A-\sigma M)^{-1}M u = \lambda u$. If we have a zero shift is equivalent to inverting the role of the mass and stiffness matrix, this is the usual hand-waving solution when $M$ is singular (for example saddle point problems) which if we use a spectral transform can be viewed from a more formal point of view.

Reply:
I've made this change, however you still don't get a particularly nice system, because of the 0 eigenvalues in the inverted system. If you want to see this, change the test to not shift the eigenvalues and you'll see that the error in the non-boundary eigenvalues increases by a lot.